lib/command/src/Obelisk/Command/Run.hs

```diff
@@ -70,37 +74,67 @@ data CabalPackageInfo = CabalPackageInfo
 preprocessorIdentifier :: String
 preprocessorIdentifier = "__preprocessor-apply-packages"
 
-run :: MonadObelisk m => m ()
+-- | Imports needed to use Obelisk.Run
+obRunImports :: [String]
+obRunImports = [ "import qualified Obelisk.Run"
+               , "import qualified Frontend"
+               , "import qualified Backend" ]
+
+profile
+  :: MonadObelisk m
+  => FilePath
+  -> m ()
+profile profileBaseName = withProjectRoot "." $ \root -> do
+  freePort <- getFreePort
+  assets <- findProjectAssets root
+  putLog Debug $ "Assets impurely loaded from: " <> assets
+  putLog Debug "Using profiled build of project."
+  liftIO $ createDirectoryIfMissing False $ takeDirectory profileBaseName
+  let -- Sane flags to enable by default, enable time profiling +
+      -- closure heap profiling.
+      rtsFlags = [ "+RTS", "-p", "-po" <> profileBaseName, "-hc", "-RTS" ]
+      nixBuildExpr = unlines $
+        [ "(with (import ./. {});"
+        , "let"
+        , "  exeSource = obelisk.nixpkgs.writeText \"ob-run\" ''"
+        , "module Main where"
+        , "import Control.Exception"
+        , "import Reflex.Profiled"
+        , "import System.Environment" ]
+        <> obRunImports <>
+        [ "main :: IO ()"
+        , "main = getArgs >>= \\args -> (Obelisk.Run.run (read (args !! 0)) (Obelisk.Run.runServeAsset (args !! 1)) Backend.backend Frontend.frontend) `finally` writeProfilingData \"" <> profileBaseName <> ".rprof\""
+        , "  '';"
+        , "in obelisk.nixpkgs.runCommand \"ob-run\" {"
+        , "  buildInputs = [ (profiled.ghc.ghcWithPackages (p: [ p.backend p.frontend])) ];"
+        , "} \"ghc -x hs -prof -fno-prof-auto ${exeSource} -o $out\")" ]
+  exePath <- nixCmd $ NixCmd_Build $ def
+      Lens.& nixBuildConfig_outLink Lens..~ OutLink_None
+      Lens.& nixCmdConfig_target Lens..~ Target
+        { _target_path = Nothing
+        , _target_attr = Nothing
+        , _target_expr = Just nixBuildExpr }
+  (_, _, _, ph) <- createProcess_ "runProfExe" $ setCwd (Just root) $ setDelegateCtlc True $ proc exePath ([show freePort, T.unpack assets] <> rtsFlags)
+  _ <- liftIO $ waitForProcess ph
+  pure ()
+
+run
+  :: MonadObelisk m
+  => m ()
 run = withProjectRoot "." $ \root -> do
   pkgs <- fmap toList . parsePackagesOrFail =<< getLocalPkgs root
-  withGhciScript pkgs root $ \dotGhciPath -> do
-    assets <- do
-      isDerivation <- readProcessAndLogStderr Debug $ setCwd (Just root) $
-        proc nixExePath
-          [ "eval"
-          , "(let a = import ./. {}; in toString (a.reflex.nixpkgs.lib.isDerivation a.passthru.staticFilesImpure))"
-          , "--raw"
-          -- `--raw` is not available with old nix-instantiate. It drops quotation
-          -- marks and trailing newline, so is very convenient for shelling out.
+  freePort <- getFreePort
```

Comment:
I made a fix to `develop` for this that we should preserve here. The call to `getFreePort` is definitely a race condition waiting to happen so we need to *at least* run it *immediately* before binding on it.